src/rgw/driver/d4n/d4n_policy.cc

```diff
@@ -473,8 +536,9 @@ void LFUDAPolicy::cleaning(const DoutPrefixProvider* dpp)
     ldpp_dout(dpp, 10) << __LINE__ << " " << __func__ << "(): e->bucket_id=" << e->bucket_id << dendl;
     ldpp_dout(dpp, 10) << __LINE__ << " " << __func__ << "(): e->user=" << e->user << dendl;
     ldpp_dout(dpp, 10) << __LINE__ << " " << __func__ << "(): e->obj_key=" << e->obj_key << dendl;
-    l.unlock();
```

Comment:
Do not remove the unlock here. Before this unlock check if the state is invalid and set a flag and then after the unlock, based on the flag, call the method to delete the data blocks and erase_dirty_object(). erase_dirty_object() internally locks and unlocks correctly. This should happen before the check for expiration time - like I mentioned before we do not need to wait for invalid entries to expire for them to be removed.

Once the entry expires then take the lock and set the state to IN_PROGRESS and release the lock. And then the other steps can be done as before